### PR TITLE
Functions Refactor

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?

This PR is for the upcoming functions refactor introducing the new execution model to allow for significantly faster execution of functions and syncronous executions.

## Test Plan

New tests have been written for syncronous execution and current tests have been updated for the new model.

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:499]:
```suggestion
    ->param('deploy', false, new Boolean(true), 'Automatically deploy the function when it is finished building.', false)
```

Seems more consistent

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:571]: Why are we giving here role:all?

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:877]: Could be very risky, any reason for the change?

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:1059]:
```suggestion
    ->desc('List Builds')
```

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:1090]: There is no search field in the data structure, do we have tests for this?

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:770]: Should be moved to a worker, no need to wait for response. We might have to introduce a `deleting` status to prevent conflicts while being deleted.

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:679]: Again, I think we can skip this output and include it in the console from a second API call.

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:1057]: Shouldn't build be nested under their tag? They should probably also use the function read and write permissions and not the execution ones. (`execution.read` -> `functions.read` and `execution.write` -> `functions.write`)

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:748]: When we delete a tag we should delete all builds

**Review by eldadfux — CHANGES_REQUESTED** [review]
Gave an initial review, added some more comments and fixes in this PR: https://github.com/appwrite/appwrite/pull/2673

File: docker-compose.yml
Comment (eldadfux) [inline docker-compose.yml:367]: We should create an executable file on our Dockerfile, like we do for other services, same for the `compose.phtml` file.

File: app/config/collections.php
Comment (lohanidamodar) [reply app/config/collections.php:2119]: @PineappleIOnic can we add functionId to builds collection, we would want to aggregate build time per function, and I'll need functionId here to do that?

File: app/config/collections.php
Comment (PineappleIOnic) [reply app/config/collections.php:1941]: This change was not required however without it the request may be confusing as we are no longer using full commands to launch servers. Instead we are directly importing the user's function file.

Comment (PineappleIOnic) [reply app/config/collections.php:1941]: This change is breaking however most of the functions refactor itself is breaking as old functions will cease to work due to the new import based functions system.

File: CONTRIBUTING.md
Comment (PineappleIOnic) [reply CONTRIBUTING.md:285]: We do have a contribution guide which explains this currently in a PR. https://github.com/appwrite/runtimes/pull/27

File: app/config/collections.php
Comment (PineappleIOnic) [reply app/config/collections.php:2119]: I'm not sure if we can add functionId to builds collection as it's supposed to stay detached from functions so it can be reused for static hosting.

File: docs/references/functions/list-builds.md
Comment (PineappleIOnic) [reply docs/references/functions/list-builds.md:1]: I'm currently 50/50 whether to keep it included, I originally added it for debugging but at the same time it could help with adding build status to the dashboard without me needing to clone the latest build information on the tag.

File: app/controllers/api/functions.php
Comment (PineappleIOnic) [reply app/controllers/api/functions.php:1057]: Builds are not currently nested in their tag since builds need to be seperate from functions so static websites can use them aswell.

File: docker-compose.yml
Comment (PineappleIOnic) [reply docker-compose.yml]: Done 👍

File: app/controllers/api/functions.php
Comment (PineappleIOnic) [reply app/controllers/api/functions.php:1090]: not yet but I'll write some

Comment (PineappleIOnic) [reply app/controllers/api/functions.php:1090]: not currently, but I can write some

File: src/Appwrite/Utopia/Response/Model/Tag.php
Comment (christyjacob4) [reply src/Appwrite/Utopia/Response/Model/Tag.php:60]: Also remove the `buildStatus` attribute from the `tags` collection

File: app/controllers/api/functions.php
Comment (PineappleIOnic) [reply app/controllers/api/functions.php]: Tests have been written.

File: app/config/collections.php
Comment (lohanidamodar) [reply app/config/collections.php]: @PineappleIOnic Then  can we sync the buildTime to the tags collection, like we are syncing other attributes? That way I can use the tags collection which already has the functionId right?

File: app/executor.php
Comment (PineappleIOnic) [reply app/executor.php:296]: I'll start looking into this now.

File: app/config/collections.old.php
Comment (lohanidamodar) [inline app/config/collections.old.php:1325]: Do we need these changes in the old collections?

File: app/config/collections.php
Comment (lohanidamodar) [inline app/config/collections.php:1901]: Make sure to update usage worker that calculates the storage taken by tags

File: app/config/collections.php
Comment (lohanidamodar) [reply app/config/collections.php:2141]: Not sure what sourceType is for?

File: app/views/install/compose.phtml
Comment (lohanidamodar) [inline app/views/install/compose.phtml:169]: do we need this?

File: app/views/install/compose.phtml
Comment (lohanidamodar) [inline app/views/install/compose.phtml:343]: why do I see two `appwrite-executor`?

File: app/tasks/usage.php
Comment (lohanidamodar) [inline app/tasks/usage.php:373]: We also need to add build sizes as we are also keeping the built code

File: app/workers/deletes.php
Comment (lohanidamodar) [inline app/workers/deletes.php:312]: We need to handle delete build

File: docs/references/functions/list-builds.md
Comment (lohanidamodar) [reply docs/references/functions/list-builds.md:1]: I say we keep it

**Review by lohanidamodar — COMMENTED** [review]
Another round of brief review

File: docs/references/functions/list-builds.md
Comment (eldadfux) [reply docs/references/functions/list-builds.md:1]: Agree, but we have to document them.

File: Dockerfile
Comment (eldadfux) [inline Dockerfile:187]: In the `docker-compose.yml` file we already have a network called `runtimes`, what is the difference between them? Can we user a term that will help differentiate between the two?

File: Dockerfile
Comment (eldadfux) [inline Dockerfile:286]: Let's move `executor` to the upper part of the list so it feels more organized. Also does it mean we need executable for `realtime` and `http`?

File: app/config/collections.php
Comment (eldadfux) [reply app/config/collections.php:1760]: +1

File: app/config/collections.php
Comment (eldadfux) [inline app/config/collections.php:1844]: What is metadata and what is it used for?

File: app/config/collections.php
Comment (eldadfux) [reply app/config/collections.php:2029]: cc @christyjacob4

File: src/Appwrite/Utopia/Response/Model/Func.php
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/Func.php:125]: Docblock is wrong here.

Comment (eldadfux) [reply src/Appwrite/Utopia/Response/Model/Func.php:125]: It would also be great to document what this filter is trying to do. (Fixing PHP default value to be {} instead of [])

File: src/Appwrite/Utopia/Response/Model/SyncExecution.php
Comment (eldadfux) [reply src/Appwrite/Utopia/Response/Model/SyncExecution.php:1]: Ping

File: src/Executor/Executor.php
Comment (eldadfux) [inline src/Executor/Executor.php:99]: Docblock is missing. Extra description will important here.

File: src/Executor/Executor.php
Comment (eldadfux) [inline src/Executor/Executor.php:78]: A lot of docblocks are missing in this class. Let's keep consistency.

File: tests/resources/docker/docker-compose.yml
Comment (eldadfux) [inline tests/resources/docker/docker-compose.yml:270]: We don't really need this here. This file is only used to test our docker compose parser class.

File: tests/resources/functions/node/index.js
Comment (eldadfux) [inline tests/resources/functions/node/index.js:13]: Would we like to test functions that also have dependencies?

File: src/Appwrite/Utopia/Response/Model/Build.php
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/Build.php:31]: If we need this docblock for maintainers, we should definitely have it for devs. `description` is automatically used in the generated model docs.

File: docker-compose.yml
Comment (eldadfux) [inline docker-compose.yml:434]: Wan't most of this settings need to use `OPENRUNTIMES` as a prefix at some point?

File: app/controllers/api/projects.php
Comment (eldadfux) [inline app/controllers/api/projects.php:604]: This is not the only place we need to update. Search for `$search =` on all files.

File: app/controllers/shared/api.php
Comment (eldadfux) [inline app/controllers/shared/api.php:79]: Why did we change this condition?

File: app/views/install/compose.phtml
Comment (eldadfux) [reply app/views/install/compose.phtml:170]: Yes, but we have a future task for refactoring all the functions mounts names.

File: app/views/install/compose.phtml
Comment (eldadfux) [inline app/views/install/compose.phtml:186]: This block is not synced with the main compose file and uses a lot of un-required env vars.

**Review by eldadfux — CHANGES_REQUESTED** [review]
@lohanidamodar it would be great if you could confirm important build, functions, execution stats are being reported as expected.

File: app/workers/builds.php
Comment (lohanidamodar) [inline app/workers/builds.php:130]: Why are we not sending entrypoint here? createRuntime requires entrypoint which is used by runtimes.

File: src/Executor/Executor.php
Comment (lohanidamodar) [inline src/Executor/Executor.php:35]: function id is never used?

**Review by lohanidamodar — COMMENTED** [review]
Initial review. Also missing entrypoint is causing errors in dart runtime

File: Dockerfile
Comment (PineappleIOnic) [reply Dockerfile:187]: They are the same network, the `appwrite_` prefix is because it's apart of the appwrite docker project

File: app/config/collections.php
Comment (PineappleIOnic) [reply app/config/collections.php:2029]: sourceType is supposed to remember what storage device was used when storing the file where as source remembers the path. You can see what it will contain here:
https://github.com/appwrite/appwrite/blob/5d9a9f53d82d464b4e1730d9883af97b562764d9/app/workers/builds.php#L93-L94

File: src/Appwrite/Utopia/Response/Model/SyncExecution.php
Comment (PineappleIOnic) [reply src/Appwrite/Utopia/Response/Model/SyncExecution.php:1]: We could merge this into one, without much problem. This was a holdover from when I wasn't sure just how different the synchronous execution model was going to be.

File: app/controllers/shared/api.php
Comment (PineappleIOnic) [reply app/controllers/shared/api.php:79]: The abuse check was still technically running the DB calls even if the _APP_OPTIONS_ABUSE environment variable was set by placing $abuse->check() after the env check we prevent that from happening

File: app/executor.php
Comment (lohanidamodar) [inline app/executor.php:205]: Some linux setups have issues while providing `CPU` and `SWAP` option, haven't encountered similar thing for Memory but it might as well cause issue, so there must be a way to disable this options - easiest was , if the environment variable is empty the we don't provide default limit. Recently got this on a digitalocean droplet

When a swap limit is set to something like 2000
```
Docker Error: WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
```
(this seems to be only warning but seems to cause build to fail)

When I set env for swap empty
```
Utopia\Orchestration\Orchestration::setSwap(): Argument #1 ($mb) must be of type int, string given, called in /usr/src/code/app/executor.php on line 205
```

File: .env
Comment (eldadfux) [reply .env:53]: I agree, we can actually do something very similar to what we do here, and keep everything consistent with the Appwrite standard. Very nice decoupling.

https://github.com/appwrite/appwrite/blob/master/docker-compose.yml#L475

File: app/controllers/shared/api.php
Comment (TorstenDittmann) [reply app/controllers/shared/api.php:79]: For performance, basically it will not write to the abuse table if abuse is disabled.

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**
